src/invert/laplace/impls/pcr/pcr.cxx

```diff
@@ -114,285 +157,732 @@ void transpose(Matrix<dcomplex>& m_t, const Matrix<dcomplex>& m) {
+  if ((global_flags & INVERT_BOTH_BNDRY_ONE) || (localmesh->xstart < 2)) {
+    inbndry = outbndry = 1;
+  }
+  if (inner_boundary_flags & INVERT_BNDRY_ONE)
+    inbndry = 1;
+  if (outer_boundary_flags & INVERT_BNDRY_ONE)
+    outbndry = 1;
+
+  int nx = xe - xs + 1; // Number of X points on this processor
+
+  // Get range of Y indices
+  int ys = localmesh->ystart, ye = localmesh->yend;
+
+  if (localmesh->hasBndryLowerY()) {
+    if (include_yguards)
+      ys = 0; // Mesh contains a lower boundary and we are solving in the guard cells
+
+    ys += extra_yguards_lower;
+  }
+  if (localmesh->hasBndryUpperY()) {
+    if (include_yguards)
+      ye = localmesh->LocalNy -
+           1; // Contains upper boundary and we are solving in the guard cells
+
+    ye -= extra_yguards_upper;
+  }
+
+  const int ny = (ye - ys + 1); // Number of Y points
+  nsys = nmode * ny;  // Number of systems of equations to solve
+  const int nxny = nx * ny;     // Number of points in X-Y
+
+  auto a3D = Matrix<dcomplex>(nsys, nx);
+  auto b3D = Matrix<dcomplex>(nsys, nx);
+  auto c3D = Matrix<dcomplex>(nsys, nx);
+
+  auto xcmplx3D = Matrix<dcomplex>(nsys, nx);
+  auto bcmplx3D = Matrix<dcomplex>(nsys, nx);
+
+  output.write("LaplaceCyclic::solve before coefs\n");
+  if (dst) {
+    output.write("LaplaceCyclic::solve in DST\n");
+    BOUT_OMP(parallel) {
+      /// Create a local thread-scope working array
+      auto k1d =
+          Array<dcomplex>(localmesh->LocalNz); // ZFFT routine expects input of this length
+
+      // Loop over X and Y indices, including boundaries but not guard cells.
+      // (unless periodic in x)
+      BOUT_OMP(for)
+      for (int ind = 0; ind < nxny; ++ind) {
+        // ind = (ix - xs)*(ye - ys + 1) + (iy - ys)
+        int ix = xs + ind / ny;
+        int iy = ys + ind % ny;
+
+        // Take DST in Z direction and put result in k1d
+
+        if (((ix < inbndry) && (inner_boundary_flags & INVERT_SET) && localmesh->firstX()) ||
+            ((localmesh->LocalNx - ix - 1 < outbndry) && (outer_boundary_flags & INVERT_SET) &&
+             localmesh->lastX())) {
+          // Use the values in x0 in the boundary
+          DST(x0(ix, iy) + 1, localmesh->LocalNz - 2, std::begin(k1d));
+        } else {
+          DST(rhs(ix, iy) + 1, localmesh->LocalNz - 2, std::begin(k1d));
+        }
+
+        // Copy into array, transposing so kz is first index
+        for (int kz = 0; kz < nmode; kz++) {
+          bcmplx3D((iy - ys) * nmode + kz, ix - xs) = k1d[kz];
+        }
+      }
+
+      // Get elements of the tridiagonal matrix
+      // including boundary conditions
+      BOUT_OMP(for nowait)
+      for (int ind = 0; ind < nsys; ind++) {
+        // ind = (iy - ys) * nmode + kz
+        int iy = ys + ind / nmode;
+        int kz = ind % nmode;
+
+        BoutReal zlen = coords->dz * (localmesh->LocalNz - 3);
+        BoutReal kwave =
+            kz * 2.0 * PI / (2. * zlen); // wave number is 1/[rad]; DST has extra 2.
```

Comment:
It sounds like "DST has extra 2" means that there are two extra cells, but why is then `nz - 3` used?